Add episode on how to write your own diagnostics (#185)
ESMValGroup/ESMValTool_Tutorial: _episodes/10-diagnostics.md

---
title: "Writing your own diagnostic script"
teaching: TBD
exercises: TBD

questions:

> - "How to read preprocessor output from a Python diagnsotics?"
> - "How to write a new diagnostic in ESMValTool?"

Reviewer: ```suggestion
- "How do I write a new diagnostic in ESMValTool?"
- "How do I read preprocessor output in a Python diagnostic?"

```

Reviewer: Switched order, changes style to match other episodes

objectives:

> - "Understanding the interface between the ESMValCore preprocessor and a diagnostic script."
> - "Writing a new Python diagnostic"

Reviewer: ```suggestion
- "Writing a new Python diagnostic"
- "Understanding the interface between the ESMValCore preprocessor and a diagnostic script."
```

Reviewer: Match order with previous suggestion.

keypoints:
- "Take home message 1"
- "etc ..."
---

## Introduction

> The diagnostic script is an important component of ESMValTool where the actual 

Reviewer: ```suggestion
The diagnostic script is an important component of ESMValTool where the
```

Reviewer: I would leave this out, as quite a lot of "science" is done in the preprocessors now.

scientific analysis or performance metric is implemented. With ESMValTool, you 
can reuse an existing diagnostic, adapt and existing one for your needs or

> write your own new diagnostic.  Diagnostics can be written in one or more open 

Reviewer: ```suggestion
write your own new diagnostic.  Diagnostics can be written in a number of open 
```

Reviewer: A single diagnostic is not usually written in multiple languages.

> source languages such as Python, R, Julia or NCL but we will focus on understanding 

Reviewer: ```suggestion
source languages such as Python, R, Julia and NCL but we will focus on understanding 
```

and writing Python diagnostics in this lesson. In order to access existing diagnostics or
to write your own, please install ESMValTool in the development mode on your 

> machine using the instructions from [here](https://esmvalgroup.github.io/ESMValTool_Tutorial/08-development-setup/index.html).

Reviewer: ```suggestion
machine using the instructions from [this episode](08-development-setup/index.html).
```

Reviewer: If it works a relative link would be better, I think.

Reviewer: also added a hint that the information is in this tutorial rather than an external site.

>
>    # Example of how to loop over variables/datasets in alphabetical order
>    groups = group_metadata(input_data, 'variable_group', sort='dataset')
>    for group_name in groups:
>        logger.info("Processing variable %s", group_name)
>        for attributes in groups[group_name]:
>            logger.info("Processing dataset %s", attributes['dataset'])
>            input_file = attributes['filename']
>            cube = compute_diagnostic(input_file)
>
>            output_basename = Path(input_file).stem
>            if group_name != attributes['short_name']:
>                output_basename = group_name + '_' + output_basename
>            provenance_record = get_provenance_record(
>                attributes, ancestor_files=[input_file])
>            plot_diagnostic(cube, output_basename, provenance_record, cfg)
>
>
>if __name__ == '__main__':
>
>    with run_diagnostic() as config:
>        main(config)
>
>~~~
>{: .language-python}
>
{:.solution}

> ## What is the starting point of the diagnostic?
>
> Can you spot the main function in the Python code above? How many references 
>do you see to this function?
>
> > ## Answer
> >

> > > The main function is defined in the middle of this script and is called near the very

Reviewer: As you probably also noticed this needs a reference to a line number :-)

Reviewer 2: Yes..I am not sure how best to do this. I will complete most of the content and revisit this.

>
>> ## Answer
>>
>> 1. One example of ``settings.yml`` can be found in the directory:
>> *path_to_recipe_output/run/map/script1/settings.yml*
>> 2. The ``metadata.yml`` files hold information
>> about the preprocessed data. There is one file for each variable having
>> detailed information on your data including project (e.g., CMIP6, CMIP5),
>> dataset names (e.g., BCC-ESM1, CanESM2), variable attributes (e.g.,
>> standard_name, units), preprocessor applied and time range of the data. You
>> can use all of this information in your own diagnostic.
> >
> >
> {: .solution}
{: .challenge}

## Diagnostic shared functions

Looking at the code in  ``diagnostic.py``, we see that ``input_data`` is
read from the ``cfg`` dictionary (line 69). Now we can group the ``input_data``
according to some criteria such as the model or experiment. To do so,
ESMValTool provides many functions such as ``select_metadata`` (line 72),
``sorted_metadata`` (line 76), and ``group_metadata`` (line 80). As you can see
in line 8, these functions are imported from ``esmvaltool.diag_scripts.shared``
that means these are shared across several diagnostics scripts. A list of

> available functions and their description can be found in [Shared diagnostic
> script code][shared].

Reviewer: ```suggestion
available functions and their description can be found in [The ESMValTool Diagnostic API reference][shared].
```

    logger.debug("Loading %s", filename)
    cube = iris.load_cube(filename)

    logger.debug("Running example computation")
    cube = iris.util.squeeze(cube)

    # Calculate a bias using the average of data
    cube.data = cube.core_data() - cube.data.mean()
    return cube
~~~

> ## iris cubes
>
> Iris reads data from NetCDF files into data structures called
> [cubes](https://scitools-iris.readthedocs.io/en/latest/userguide/iris_cubes.html).
> The data in these cubes can be modified, combined with other cubes' data or
> plotted.
{: .callout}

> ## Reading data using xarray
>
> Alternately, you can use [xarrays](http://xarray.pydata.org/en/stable/) to read the data
> instead of Iris.
>
>> ## Answer
>>
>> First, import ``xarray`` package at the top of the script as:
>>
>>~~~python
>>import xarray as xr
>>~~~
>>
>> Then, change the ``compute_diagnostic`` as:
>>
>>~~~python
>>def compute_diagnostic(filename):
>>    """Compute an example diagnostic."""
>>    logger.debug("Loading %s", filename)
>>    dataset = xr.open_dataset(filename)
>>
>>    #do your analyses on the data here
>>
>>    return dataset
>>~~~
>>
> {: .solution}
{: .challenge}

> > ## Reading data using Scipy's netCDF library

Reviewer: I think this is the "old" NetCDF library that does not handle netcdf4. Perhaps we should use the one from unidata as the example instead? http://unidata.github.io/netcdf4-python/

provenance. To do so, we use the function ``get_provenance_record`` (line 99).
Let us have a look at the definition of this function in line 21 where we
describe the diagnostic data and plot. Using the dictionary ``record``, it is
possible to add custom provenance to our diagnostics output. 
Provenance is stored in the *W3C PROV XML*
format and also in an *SVG* file under the ``work`` and ``plot`` directory. For
more information, see [recording provenance][provenance].

## Congratulations!

You now know the basic diagnostic script structure and some available tools for putting 
together your own diagnostics. Have a look at existing recipes and diagnostics in the 
repository for more examples of functions you can use in your diagnostics!
{% include links.md %}

[recipe]: https://github.com/ESMValGroup/ESMValTool/blob/master/esmvaltool/recipes/examples/recipe_python.yml
[diagnostic]: https://github.com/ESMValGroup/ESMValTool/blob/master/esmvaltool/diag_scripts/examples/diagnostic.py
[interface]: https://docs.esmvaltool.org/projects/esmvalcore/en/latest/interfaces.html
[shared]: https://docs.esmvaltool.org/en/latest/api/esmvaltool.diag_scripts.shared.html

> [netCDF]: https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.io.netcdf.netcdf_file.html

Reviewer: Perhaps update to a newer version?
